src/main/java/org/benetech/servicenet/matching/counter/LocationSimilarityCounter.java

```diff
@@ -33,68 +35,100 @@
+        List<GeocodingResult> partnerGeocoding) {
         float max = NO_MATCH_RATIO;
 
-        List<GeocodingResult> baseGeocoding = getGeoCodingResult(location1);
-        List<GeocodingResult> partnerGeocoding = getGeoCodingResult(location2);
         for (GeocodingResult result1 : baseGeocoding) {
             for (GeocodingResult result2 : partnerGeocoding) {
-                max = Math.max(max, countSimilarityRatio(result1, result2));
+                max = Math.max(max, countSimilarityRatio(result1.getGoogleCoords(), result2.getGoogleCoords()));
             }
         }
 
         return max;
     }
 
-    private List<GeocodingResult> getGeoCodingResult(Location location) {
-        if (coordinatesAreMissing(location)) {
-            return location.getGeocodingResults();
-        } else {
-            return singletonList(new GeocodingResult(location.getName(), location.getLatitude(), location.getLongitude()));
-        }
+    private List<GeocodingResult> getGeoCodingResult(Location location, MatchingContext context) {
+        return geocodingResultService.findAllForAddressOrFetchIfEmpty(location.getPhysicalAddress(), context);
     }
 
-    private boolean coordinatesAreMissing(Location location1) {
-        return location1.getLongitude() == null || location1.getLongitude().equals(0.0);
+    private List<GeocodingResult> mockGeocodingResultsWithProviderData(Location location) {
+        return singletonList(new GeocodingResult(location.getName(), location.getLatitude(), location.getLongitude()));
     }
 
-    private float countRatioIfCoordinatesArePresent(Location location1, Location location2) {
-        return countSimilarityRatio(
-            getStraightLineDistanceInMeters(location1.getCoordinates(), location2.getCoordinates()));
+    private boolean areCoordsNotComparable(List<GeocodingResult> baseGeocoding,
+        List<GeocodingResult> partnerGeocoding, Location baseLocation, Location partnerLocation) {
+
+        return areGeocodingResultsNotReliable(baseGeocoding) && areProviderCoordEmpty(baseLocation)
+                || areGeocodingResultsNotReliable(partnerGeocoding) && areProviderCoordEmpty(partnerLocation);
     }
 
-    private boolean anyPhysicalAddressIsNull(Location location1, Location location2) {
-        return location1 == null || location1.getPhysicalAddress() == null
-            || location2 == null || location2.getPhysicalAddress() == null;
+    private boolean areProviderCoordEmpty(Location location) {
+        return location.getLongitude() == null || location.getLatitude() == null
+            || location.getLongitude().equals(0.0) || location.getLatitude().equals(0.0);
     }
 
-    private boolean allCoordinatesArePresent(Location location1, Location location2) {
-        return !coordinatesAreMissing(location1) && !coordinatesAreMissing(location2);
+    private boolean areGeocodingResultsNotReliable(List<GeocodingResult> geocodingResults) {
```

Comment:
I think it would be better to use anyMatch and name this method "areGeocodingResultsReliable"

Reply:
Yeah, I had it like that but followed Idea suggestion to inverse the method. I think it may be more readable with anyMatch though.